File: ruby/find_first_duplicate.rb
require 'set'

def find_first_duplicate(arr)
  uniques = Set.new

  arr.each do |v|
    return v if uniques.include?(v)

    uniques.add(v)
  end

  -1

end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  puts "Expecting: 5"
  puts "=>", find_first_duplicate([1, 4, 5, 3, 5, 8, 7, 6])
end

# Initialize an empty set (sets only contain unique values). Loop through the array
# checking if the value is in the set. If so, return the value. If no duplicate
# is found, return -1.

# Sets only contain unique values. If a value in the array is in the set, it 
# is a duplicate. The first duplicate found should be returned.

File: javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  const uniques = new Set();

  for (let v of arr) {
    if (uniques.has(v)) {
      return v;
    }
    uniques.add(v);
  }
  return -1;
}

if (require.main === module) {
  
  console.log("Expecting: 5");
  console.log("=>", findFirstDuplicate([1, 4, 3, 5, 6, 7, 8, 5]));

  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;

// Initialize empty set, then loop through the array. If the set contains 
// the current value, return the current value. Else, return -1.

// Sets only contain unique values. If the set contains the current element
// of the array, then it is a duplicate. The first duplicate to be found should
// be returned.